### PR TITLE
try different GPGME libraries for Python

### PR DESCRIPTION
**Pull request by vnev**
Trying to use pygpgme as the dependency with which to move `pgpg` further.  @dashohoxha I would appreciate any input if this is the right way to move. Thanks!

### CODE REVIEW

**Comment by diveshuttam** [comment]
Just some inputs:
- We should try these out before introducing them as dependencies. 
- Also, as I suggested in one of the earlier discussion (quoted below) PYGPGME and PYME are not being developed actively. So, I would like to test the official bindings first, if they don't work then we may switch on to one of PyGPGME or PYME.
>GPGME provides [official bindings for Python](https://packages.debian.org/sid/python3-gpg), we plan to use them only. Existing wrappers like [PYME](https://github.com/adversary-org/pyme3) , and [PYGPGME](https://github.com/rshk/pygpgme) are no longer being actively developed. 
>You can find few more details regarding this in my [GSoC proposal draft](https://tinyurl.com/diveshuttam-GSoC18-proposal
).
- Also If you wanna help, you may try these bindings and give your valuable inputs. It would help us a lot. I will be starting with the same in a couple of days. 

cc: @dashohoxha, I don't think it is the best thing to do at the moment. But if @vnev is willing to help us it would be great if we can assign a few tasks to him also.

**Comment by dashohoxha** [comment]
> if @vnev is willing to help us it would be great if we can assign a few tasks to him also

If there are some suitable tasks that he can do, why not.

**Comment by vnev** [comment]
Thank you all for the prompt responses. I will get to trying to official bindings and update the PR.

**Comment by yugeshk** [comment]
@diveshuttam 
>Just some inputs:
>
>    We should try these out before introducing them as dependencies.
    Also, as I suggested in one of the earlier discussion (quoted below) PYGPGME and PYME are not being developed actively. So, I would like to test the official bindings first, if they don't work then we may switch on to one of PyGPGME or PYME.
>
>    GPGME provides official bindings for Python, we plan to use them only. Existing wrappers like PYME , and PYGPGME are no longer being actively developed.
>    You can find few more details regarding this in my GSoC proposal draft.

I read on some archive of a mailing list last night that PyMe was eventually folded into the GPGME project and is currently maintained by them under no stand-alone name. So maybe you'd like to look at PyGPGME only and not the last release of the "PyMe" which was back in 2008. (I'll share that mail as soon as I find it again!)

I too am doing some tests of my own. Probably this thread will help us all out.

**Comment by diveshuttam** [comment]
@yugeshk yes you are correct I also read it on some mailing list. AFAIR it was developed as a fork of PYME.
I had added these links to [my proposal](https://tinyurl.com/diveshuttam-GSoC18-proposal) you may have a look, it may be helpful in your project also.

**Comment by vnev** [comment]
I think so far there is extensive documentation for GPGME's official bindings, which can be found [here](http://files.au.adversary.org/crypto/GPGMEpythonHOWTOen.html). There is good explanation of what every function does as well as examples to go with it. I would recommend going with this, however if there's a reason why someone would like to use the alternative libraries then I am all ears

File: src/cmd/init.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+def main():
+    if len(sys.argv) == 0:
+        cmd_init_help()
+    print(str(sys.argv))
+    if len(sys.argv) > 2:
+        cmd_init(sys.argv[2])
+    else:
+        cmd_init()
+
+if __name__ == "__main__":
```
Comment (sourabhtk37) [inline src/cmd/init.py]: This doesn't seem right. `__init__.py` files are used to specify the directory as a package and not as main driver programs. Move this logic into main driver program (such as `main.py` or `pgpg.py`)

File: src/cmd/init.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+                shutil.rmtree(fullPath)
+                os.makedirs(fullPath)
+            else:
+                exit(1)
+        else:
+            os.makedirs(fullPath)
+
+
+def main():
+    if len(sys.argv) == 0:
```
Comment (sourabhtk37) [inline src/cmd/init.py]: Use inbuilt-argparse library to build command line options and parsing. Will help you in generating help, adding options easily and of course parsing.

File: src/cmd/init.py (hunk 1 above)
Comment (vnev) [reply src/cmd/init.py]: Yup I will do that. I just had it here for testing purposes

**Comment by vnev** [comment]
I have gone ahead and completed the basic `init` functionality. Currently the main function is within the `init.py` file itself for testing purposes. Also, as @sourabhtk37 mentioned, `argparse` can definitely help us in the long run, so I will try to implement that tomorrow. Hopefully this is acceptable for now!

**Comment by dashohoxha** [comment]
Please keep the discussion about which library is best (or pros and cons of each library) here:
 - https://github.com/orgs/EasyGnuPG/teams/egpg/discussions/5
 - https://github.com/EasyGnuPG/pgpg/issues/8

**Comment by dashohoxha** [comment]
@vnev it's Ok that you try to help, but please let @diveshuttam lead the project and do the main work, because he has been selected for this project, he is responsible for it, he will be evaluated, and finally he will get payed (if the project is successful).
So, I am not going to approve your pull request unless @diveshuttam is Ok with it.

Personally I think that you are going a bit too fast. It is a necessary requirement that you know very well the functionality and code of the existing `egpg`, before you start implementing `pgpg`. Maybe you have done it already, but I don't see any evidence of it.

One way of making sure that you understand its functionality and code is to try upgrading it to gnupg-2.2: https://github.com/EasyGnuPG/egpg/issues/27 Would you like to give it a try?

**Comment by vnev** [comment]
@dashohoxha just to clarify: you'd like me to tackle the EasyGnuPG/egpg#27 issue? Definitely I can go in and test gnupg-2.2 if that is what you're asking. I am already running Ubuntu bionic on a virtual machine with 2.2 installed, so its only a matter of testing it at this point.

**Comment by dashohoxha** [comment]
> you'd like me to tackle the EasyGnuPG/egpg#27 issue

It is not a matter of tackling, it is about you getting familiar with the functionality and code of `egpg`.

> It is up to @diveshuttam if he would like to approve or deny this.

Since he is the person that is responsible for this project, it is only fair to let him do things the way he thinks is best.

**Comment by vnev** [comment]
@dashohoxha actually I am familiar with the workings of `egpg`. However I was not aware that 2.2 had a few issues which are evident after some of the tests fail. I will go ahead and see how to improve or fix those errors and maybe update issue EasyGnuPG/egpg#27 in case I come across anything, before I go any further with this codebase

**Comment by diveshuttam** [comment]
Hi @vnev, Thanks for that link. It will help us a lot.
Also, I appreciate your efforts but this is not the direction I am planning to take this project to.
Though your approach is not wrong, but as you are using argparse, we might end up reinventing the wheel for many of the functionalities that I want to add into pgpg's interface. I plan to use [click library](http://click.pocoo.org/5/) for this purpose, it will help us in getting many of the functionalities on the go and minimize the development time.
As of now I won't be able to accept this PR. Can you wait for some time, probably around a week for me to catch up with my schedule, I'll surely add some issues in which you can help directly. 
By the time you can definitely look into https://github.com/EasyGnuPG/egpg/issues/27, I'll join you soon. Also I would really urge you to read my proposal so that we are on the same page for this project. 
Lastly, your suggestions are always welcome during this time.

**Comment by diveshuttam** [comment]
Also if you would like to work on #8 and trying other libraries, You can keep on adding the code to this (try-pygpgme) branch and add your comments to the issue page. I'll definitely get back to them before deciding on the required library.

**Comment by vnev** [comment]
@diveshuttam thanks for the valuable feedback. I will definitely check out your proposal and try to see what I can do about EasyGnuPG/egpg#27.

**Pull request closed without merging**